### ISSUE TITLE
High latency when using fetch() API

### ISSUE THREAD

**Issue by lalithsuresh** [jOOQ/jOOQ#11099]
### Expected behavior

I'm observing quite a lot of additional latency when using `fetch()` to gather a Result<Record> and wanted to understand why.

I have a custom `MockDataProvider`, where I return a cached result for queries of the form "select * from X". Each result contains 10K records with 3 columns each (a `varchar(36)`, an `integer`, and a `boolean`). In my provider, I effectively call:

```java
        final Result<Record> result = dslContext.newResult(fields);
        result.addAll(materializedViews.get(tableName.toUpperCase()));
        return new MockResult(1, result);
```

Now here are the different ways I've tried to pull this cached data out.

* First, as a baseline, I exposed a public method on my `MockDataProvider` implementation that only invokes the above logic.

```java
provider.fetchTable("hostsv");
```
This call takes about 80us on average.

* Second, I tried to use JOOQ's `fetchResultSet` API:

```java
create.selectFrom("hostsv").fetchResultSet();
```
This takes about 140us per call on average.

* Third, I tried using JOOQ's `fetch()` API:

```java
create.selectFrom("hostsv").fetch();
```
...takes about 3ms on average, which seems quite high.

Digging deeper with a profiler, I see a lot of time is being spent within `CursorImpl.fetch` and what I assume is repeated work to figure out how to translate the result record I supplied. I don't know JOOQ internals well, but I'm not sure why this conversion cost is being paid on every call (when I've already supplied a `Result<Record>`?)

![image](https://user-images.githubusercontent.com/434651/101269018-a31c0080-371e-11eb-9d1d-f70769ee367e.png)

I'm curious to know what I'm missing. Thanks!

### Actual behavior

I'd expect fetch() to be faster.

### Steps to reproduce the problem

- Create a `MockDataProvider`
- Intercept select calls and return a Result<Record> with 10K rows, each with 3 columns (a varchar(36), an integer and a boolean).
- Run `create.selectFrom(tableName).fetch()`.

### Versions

- jOOQ: tested on 3.13.1 and 3.14.5
- Java: 12.0.1
- Database (include vendor): custom, but also tested with H2
- OS:
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks for your report. May I ask you to provide your "steps to reproduce" in code, rather than a description, to make sure I'll be measuring the exact same thing when trying to reproduce this?

**Comment by lukaseder** [comment]
To explain: If I access fields by index rather than by name (string) or name (qualified `org.jooq.Name`) or by field reference (the same identity as in the result vs a different instance) might all influence the measurements...

**Comment by lalithsuresh** [comment]
@lukaseder Understood! The current code comes with some baggage. I'll share a minimal example later today.

**Comment by lalithsuresh** [comment]
@lukaseder here's a self-contained JMH benchmark. I've tried to strip away as many details as I could from the original code. I've stripped out all the SQL parsing logic. I'm able to measure the same high latency with `fetch()` I mentioned above. Performance seems to have gone from about 3ms per `fetch()` to 4ms per `fetch()` between JOOQ 3.13.1 and 3.14.4.

The key things I've retained in the provider from the original code (see [1]) are:

* The way I extract Table<?> and Field<?> information from `conn.meta()` in the `MockDataProvider` implementation's constructor
* The way I'm constructing records using the above extracted Field<?> information. That is, I do `Record record = conn.newRecord(fields)`, followed by calls to `record.setValue(field, <val>)`.

Let me know if you need more information. Thanks a lot! :)

```java
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 100, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
public class FetchBench {
    @Nullable DSLContext create = null;
    @Nullable DataProvider provider = null;

    @Setup(Level.Trial)
    public void setUpDb() {
        final String s1 = "create table hosts (id varchar(36), capacity integer, up boolean)";
        provider = new DataProvider(List.of(s1));
        final MockConnection connection = new MockConnection(provider);
        create = DSL.using(connection);
        for (int i = 0; i < 10000; i++) {
            create.execute("insert into hosts values ('n', 10, true)");
        }
    }

    @Benchmark
    public Result<Record> fetchDirect() {
        return provider.fetchTable("hosts");
    }

    @Benchmark
    public ResultSet fetchResultSet() {
        return create.selectFrom("hosts").fetchResultSet();
    }

    @Benchmark
    public Result<Record> fetch() {
        return create.selectFrom("hosts").fetch();
    }

    private static class DataProvider implements MockDataProvider {
        private final Map<String, Set<Record>> materializedViews = new ConcurrentHashMap<>();
        private final Map<String, List<Field<?>>> tablesToFields = new HashMap<>();
        private final DSLContext conn = using("jdbc:h2:mem:");
        private final AtomicInteger insertCount = new AtomicInteger(0);
        private final Field<Integer> updateCountField = field("UPDATE_COUNT", Integer.class);

        public DataProvider(final List<String> ddl) {
            ddl.forEach(conn::execute);
            for (final Table<?> table: conn.meta().getTables()) {
                final Schema schema = table.getSchema();
                assert schema != null;
                if (schema.getName().equals("PUBLIC")) { // H2-specific assumption
                    tablesToFields.put(table.getName(), Arrays.asList(table.fields()));
                }
            }
        }

        @Override
        public MockResult[] execute(final MockExecuteContext ctx) throws SQLException {
            final String sql = ctx.sql().toUpperCase();
            final MockResult[] ret = new MockResult[1];
            if (sql.startsWith("INSERT")) {
                final String tableName = "HOSTS";
                final List<Field<?>> fields = tablesToFields.get(tableName);
                final Record record = conn.newRecord(fields);

                // Not bothering to parse the input data.
                for (final Field<?> field: fields) {
                    final DataType<?> dataType = field.getDataType();
                    switch (dataType.getSQLType()) {
                        case Types.BOOLEAN:
                            record.setValue((Field<Boolean>) field, true);
                            break;
                        case Types.INTEGER:
                            record.setValue((Field<Integer>) field, 10);
                            break;
                        case Types.VARCHAR:
                            record.setValue((Field<String>) field, "n" + insertCount.incrementAndGet());
                            break;
                        default:
                            throw new RuntimeException("Unknown datatype " + field);
                    }
                }
                final Set<Record> materializedView =
                        materializedViews.computeIfAbsent(tableName, (k) -> new LinkedHashSet<>());
                materializedView.add(record);
                final Result<Record1<Integer>> result = conn.newResult(updateCountField);
                final Record1<Integer> resultRecord = conn.newRecord(updateCountField);
                resultRecord.setValue(updateCountField, 1);
                result.add(resultRecord);
                ret[0] = new MockResult(1, result);
            } else if (sql.startsWith("SELECT")) {
                final String tableName = "HOSTS";
                final Result<Record> result = fetchTable(tableName);
                ret[0] = new MockResult(1, result);
            } else {
                ret[0] = new MockResult(new SQLException("Unsupport op"));
            }
            return ret;
        }

        public Result<Record> fetchTable(final String tableName) {
            final List<Field<?>> fields = tablesToFields.get(tableName.toUpperCase());
            if (fields == null) {
                throw new RuntimeException(String.format("Unknown table %s queried", tableName));
            }
            final Result<Record> result = conn.newResult(fields);
            result.addAll(materializedViews.computeIfAbsent(tableName.toUpperCase(), (k) -> new LinkedHashSet<>()));
            return result;
        }
    }
}
```

[1] https://github.com/vmware/differential-datalog/blob/master/sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java

**Comment by lukaseder** [comment]
Thanks a lot for your benchmark. I can reproduce similar overheads with it:

![image](https://user-images.githubusercontent.com/734593/101764808-cbba4680-3ae0-11eb-8d64-6563676f8db0.png)

Benchmark result on my machine:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt  100  4324044.091 ± 48127.992  ns/op
FetchBench.fetchDirect     avgt  100    95642.217 ±   655.732  ns/op
FetchBench.fetchResultSet  avgt  100    98525.221 ±  1083.377  ns/op
```

The two benchmark methods that are supposed to use the `MockResult` don't make a meaningful comparison. `fetchResultSet()` doesn't consume the result set. This does:

```java
    @Benchmark
    public void fetchResultSet(Blackhole bh) throws SQLException {
        try (ResultSet rs = create.selectFrom("hosts").fetchResultSet()) {
            while (rs.next()) {
                bh.consume(rs.getString(1));
                bh.consume(rs.getInt(2));
                bh.consume(rs.getBoolean(3));
            }
        }
    }
```

The result is then:

```
FetchBench.fetch           avgt  100  4240004.765 ± 39621.661  ns/op
FetchBench.fetchDirect     avgt  100    83370.111 ±   944.099  ns/op
FetchBench.fetchResultSet  avgt  100  1479589.405 ± 16671.387  ns/op
```

The `fetchDirect()` method doesn't traverse anything either. It's obviously better to cache the `Result` data structure directly, rather than having jOOQ go through the `MockResultSet` API to copy things from your cached `Result` to a new `Result`. However, the difference shouldn't be as big.

I don't think the `getConverter()` method is the cause here, that's probably just a profiling artifact... The `org.jooq.tools.Convert::convert0` method is a better lead.

**Comment by lukaseder** [comment]
Related issues https://github.com/jOOQ/jOOQ/issues/5865, https://github.com/jOOQ/jOOQ/issues/6799 (probably duplicates this)

**Comment by lukaseder** [comment]
Anyway. This `MockResultSet::get` method is in a hot loop (columns x rows), so any redundant work that it can avoid should be moved out. The `getConverter()` call and the subsequent wrapping of the converter in `Converters.inverse()` can be cached when initialising the `MockResultSet`

Result of that improvement:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt  100  3532638.712 ± 23931.199  ns/op
FetchBench.fetchDirect     avgt  100    80180.138 ±   719.707  ns/op
FetchBench.fetchResultSet  avgt  100   943714.809 ± 11244.001  ns/op
```

Already quite significant. I'll commit things step by step in this issue... First, caching things (little regression risk), then refactoring things (higher regression risk)

**Comment by lukaseder** [comment]
Further improvements (see subsequent commit for details):

- Shortcut in `Convert::convert` if the value is already of the right type, prevents the `ConvertAll` allocations, and a few branches
- Shortcut in `Convert::convert0` if the `Converter.fromType()` is `Object.class`, then we don't need an additional `ConvertAll<>(fromType)` round trip, which is a no-op
- In `ConvertAll`, delay the calls to `wrapper(Class)` until they're needed
- In `ConvertAll`, shortcuts should be ordered as: 
  - null check
  - class equality check
  - class assignability check
  - wrapper type equality check

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  3204033.628 ± 17168.761  ns/op
FetchBench.fetchDirect     avgt   50    83576.796 ±   866.782  ns/op
FetchBench.fetchResultSet  avgt   50   631722.508 ±  5634.768  ns/op
```

**Comment by lukaseder** [comment]
Further improvements (in the next commit):

- Converters.inverse(IdentityConverter) should return the `IdentityConverter` itself, to enable the below improvement:
- Convert::convert0 could implement a quick instanceof check to see if we're converting things with the `IdentityConverter`, in case of which a no-op can be implemented as a short cut.

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  3138709.956 ± 44575.893  ns/op
FetchBench.fetchDirect     avgt   50    80511.832 ±  1145.242  ns/op
FetchBench.fetchResultSet  avgt   50   486674.536 ±  7207.628  ns/op
```

**Comment by lukaseder** [comment]
Note that some `RecordImpl[N]` constructor performance fixes are done via a different task: https://github.com/jOOQ/jOOQ/issues/11058

They might play into measurements as well

**Comment by lukaseder** [comment]
An interesting observation is the call to `getAttachables()`, which accounts for 3% in your profiling session, or 5% after my recent improvements here:

![image](https://user-images.githubusercontent.com/734593/101781161-c3213a80-3af7-11eb-995f-9d6f0f71f056.png)

This loop rarely needs to be performed, as few data types are attachable (nested records, such as UDTs are). But we're looping for *every record*, instead of only once for the result (in the negative case). I've created a separate issue for this: https://github.com/jOOQ/jOOQ/issues/11111

**Comment by lukaseder** [comment]
No significant improvement can be observed:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  3101229.310 ± 54741.724  ns/op
FetchBench.fetchDirect     avgt   50    93851.358 ±  1334.669  ns/op
FetchBench.fetchResultSet  avgt   50   507123.521 ±  9454.959  ns/op
```

Seems to be a profiling artifact...

**Comment by lukaseder** [comment]
*But*. There's another place where we try to attach things in `DefaultBinding.AbstractBinding::attach`. To my utter surprise, removing that `attach` call yields *much* better results:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  2285512.763 ± 20108.728  ns/op
FetchBench.fetchDirect     avgt   50    80354.515 ±  1092.744  ns/op
FetchBench.fetchResultSet  avgt   50   477516.160 ±  5394.259  ns/op
```

The problem wasn't really visible in the profiler as the branch was cut off:

**Before**

![image](https://user-images.githubusercontent.com/734593/101783465-99b5de00-3afa-11eb-9cdb-7181459d3bdc.png)

**After**

![image](https://user-images.githubusercontent.com/734593/101783531-ad614480-3afa-11eb-92fc-d0e732183148.png)

I feel the profiler is lying to me again (as ever so often). Perhaps, avoiding that `attach` call simply benefits some inlining threshold?

**Comment by lukaseder** [comment]
Caching attachability based on a declared data type of a `Binding.converter().toType()` is risky, because the `Converter::toType` is not required to declare being a subtype of `Attachable`. E.g., a `Converter<T, Object>` could still produce an instance of `Attachable`. Seems like a risky improvement...

**Comment by lukaseder** [comment]
But we can implement the improvement for all the types in `SQLDataType`, which implies no custom converter is being used.

**Comment by lukaseder** [comment]
Still better with the improvement:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  2431977.577 ± 65139.614  ns/op
FetchBench.fetchDirect     avgt   50    94646.123 ±  1608.051  ns/op
FetchBench.fetchResultSet  avgt   50   494055.116 ± 14119.938  ns/op
```

And this test still passes, where attachables are not required to be declared by converters:

```java
static interface I {
    Integer i();
    Configuration configuration();
}

static class C implements Attachable, I {
    final Integer i;
    Configuration c;

    C(Integer i) {
        this.i = i;
    }

    @Override
    public Integer i() {
        return i;
    }

    @Override
    public void attach(Configuration configuration) {
        this.c = configuration;
    }

    @Override
    public void detach() {
        this.c = null;
    }

    @Override
    public Configuration configuration() {
        return c;
    }
}

@Test
public void testAttachables() {
    Field<I> i = field("ID1", INTEGER.asConvertedDataType(I.class, C::new, I::i));
    Field<C> c = field("ID1", INTEGER.asConvertedDataType(C.class, C::new, C::i));

    DSLContext e = DSL.using(new MockConnection(Mock.of(recordOne)), SQLDialect.H2);
    I vi = e.fetchValue(i);
    C vc = e.fetchValue(c);

    assertEquals(1, vi.i());
    assertNotNull(vi.configuration());
    assertEquals(1, vc.i());
    assertNotNull(vc.configuration());
}
```

**Comment by lukaseder** [comment]
While I'm at it, let's drop sampling and look into an actual profiling session.

The top 3 allocated objects are:

- The `RecordImpl3` types (inevitable)
- The `CursorRecordInitialiser` (used to abstract over nesting records)
- The `RecordDelegate` (used to abstract over the `RecordListener` SPI)

![image](https://user-images.githubusercontent.com/734593/101791015-5b70ec80-3b03-11eb-8b1f-ab7cf361ec4c.png)

We might be able to recycle the latter two per `Cursor` usage, which would mean 10000 less allocations for a result with 10000 rows. Could be worth it. Will explore.

**Comment by lukaseder** [comment]
Preventing the `CursorRecordInitialiser` re-allocation is simple and leads to a small improvement:

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  2257800.995 ± 46440.720  ns/op
FetchBench.fetchDirect     avgt   50    80113.985 ±  1164.623  ns/op
FetchBench.fetchResultSet  avgt   50   499842.915 ± 12986.001  ns/op
```

**Comment by lukaseder** [comment]
Avoiding the `RecordDelegate` allocation doesn't add much anymore. It's still better without the allocation, though.

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  2355472.426 ± 94835.165  ns/op
FetchBench.fetchDirect     avgt   50    94459.598 ±  1115.137  ns/op
FetchBench.fetchResultSet  avgt   50   485924.534 ±  4528.558  ns/op
```

**Comment by lalithsuresh** [comment]
Thanks for the steady stream of updates. :) It's also nice that fetchResultSet() seems to be getting faster with your changes.

**Comment by lukaseder** [comment]
> Thanks for the steady stream of updates. :) It's also nice that fetchResultSet() seems to be getting faster with your changes.

The logic inside `MockResultSet` is the same, so both these benchmarks will profit.

Of course, ultimately, the one thing that cannot be optimised here is the SQL string generation and allocation of new records and result arraylist. So, if you *know* that you can cache the result on a higher level than from below the JDBC API, that will always perform better. In other words, `fetchDirect` will remain faster than `fetchResultSet` and `fetch`, irrespective of further improvements.

**Comment by lukaseder** [comment]
Do you have time to re-run your benchmarks on 3.14.5-SNAPSHOT? I've pushed the improvements so far to both 3.15.0-SNAPSHOT and 3.14.5-SNAPSHOT branches... They'll also be included in the next nightly builds available from here, if you're licensed: https://www.jooq.org/download/versions

**Comment by lalithsuresh** [comment]
I can try to re-run the benchmarks, yes.

**Comment by lukaseder** [comment]
There. I found another potential improvement. If we cache the current record in the `MockResultSet` rather than accessing `result.get(index - 1)` all the time, we can slightly speed things up once more

```
Benchmark                  Mode  Cnt        Score       Error  Units
FetchBench.fetch           avgt   50  2187359.979 ± 29736.155  ns/op
FetchBench.fetchDirect     avgt   50    84973.816 ±  2816.025  ns/op
FetchBench.fetchResultSet  avgt   50   452909.908 ±  5869.055  ns/op
```

**Comment by lalithsuresh** [comment]
> > Thanks for the steady stream of updates. :) It's also nice that fetchResultSet() seems to be getting faster with your changes.
> 
> The logic inside `MockResultSet` is the same, so both these benchmarks will profit.
> 
> Of course, ultimately, the one thing that cannot be optimised here is the SQL string generation and allocation of new records and result arraylist. So, if you _know_ that you can cache the result on a higher level than from below the JDBC API, that will always perform better. In other words, `fetchDirect` will remain faster than `fetchResultSet` and `fetch`, irrespective of further improvements.

@lukaseder Yes, I'm able to reproduce the measurements with 3.15.0-SNAPSHOT and they look similar to what you found. Thanks!

I expect `fetchDirect()` to be faster as well. My goal is to make an in-memory computation engine have an SQL interface, and have code that's already using JOOQ to benefit from it. Is there a way in the JOOQ API to cache this data further up the call chain, but still behind a DSLContext?

That is, ideally, the `Result<Record>` won't need to be converted to a JDBC `ResultSet` and then back again to `Result<Record>` by JOOQ. At least I suspect that's what is going on right now, with all the `Convert`-related calls.

**Comment by lukaseder** [comment]
> Yes, I'm able to reproduce the measurements with 3.15.0-SNAPSHOT and they look similar to what you found. Thanks!

Great to hear! Thanks for the confirmation. I'll chase this a bit more next week.

> Is there a way in the JOOQ API to cache this data further up the call chain, but still behind a DSLContext?

Not yet, see: https://github.com/jOOQ/jOOQ/issues/2752, https://github.com/jOOQ/jOOQ/issues/2751. But even then, you'd *still* have to build your dynamic jOOQ query and generate the SQL string (with inlined bind values!) before you can make a cache hit/miss decision. I would think that implementing a cache on the DAO level on your side should be much simpler, no?

> That is, ideally, the cached `Result<Record>` won't need to be converted to a JDBC `ResultSet` and then back again to `Result<Record>` by JOOQ. At least I suspect that's what is going on right now, with all the `Convert`-related calls.

Yes, the point of the JDBC mocking APIs is not to bypass JDBC but to replace things on the JDBC level. Perhaps, the jOOQ `Cursor` could check the `ResultSet` and if it's a `MockResultSet`, extract its result directly to avoid *some* of the `Convert` calls. But it could still not trust the contents. Your `MockResultSet` might produce more columns, or different data types (e.g. `Long` instead of `Integer`), so we'd still have to make `Convert` calls to make sure the expected data type is produced, and wrap data in new record instances for a variety of reasons. I don't think we'd win too much, yet complicate a lot of things.

**Comment by lalithsuresh** [comment]
> > Yes, I'm able to reproduce the measurements with 3.15.0-SNAPSHOT and they look similar to what you found. Thanks!
> 
> Great to hear! Thanks for the confirmation. I'll chase this a bit more next week.
> 
> > Is there a way in the JOOQ API to cache this data further up the call chain, but still behind a DSLContext?
> 
> Not yet, see: #2752, #2751. But even then, you'd _still_ have to build your dynamic jOOQ query and generate the SQL string (with inlined bind values!) before you can make a cache hit/miss decision. I would think that implementing a cache on the DAO level on your side should be much simpler, no?
> 

Looking forward to these changes! So far, creating the dynamic JOOQ query hasn't been much of an overhead. We may not able to cache at the DAO level (because we don't know the schema in advance), but we should be able to expose `Result<Record>` further up the chain. 

> > That is, ideally, the cached `Result<Record>` won't need to be converted to a JDBC `ResultSet` and then back again to `Result<Record>` by JOOQ. At least I suspect that's what is going on right now, with all the `Convert`-related calls.
> 
> Yes, the point of the JDBC mocking APIs is not to bypass JDBC but to replace things on the JDBC level. Perhaps, the jOOQ `Cursor` could check the `ResultSet` and if it's a `MockResultSet`, extract its result directly to avoid _some_ of the `Convert` calls. But it could still not trust the contents. Your `MockResultSet` might produce more columns, or different data types (e.g. `Long` instead of `Integer`), so we'd still have to make `Convert` calls to make sure the expected data type is produced, and wrap data in new record instances for a variety of reasons. I don't think we'd win too much, yet complicate a lot of things.

Yup, understood! The added complexity isn't warranted in this case.

**Comment by lukaseder** [comment]
Hm, I thought this issue was closed already. I've had another look at the profile. On a different machine, I'm getting these results for the benchmark:

```
Benchmark                  Mode  Cnt        Score      Error  Units
FetchBench.fetch           avgt   50  1504071.362 ± 8495.837  ns/op
FetchBench.fetchDirect     avgt   50    29403.693 ±  462.115  ns/op
FetchBench.fetchResultSet  avgt   50   334022.352 ± 1882.968  ns/op
```

There's probably still more that can be done, e.g. avoiding the `ConverterProvider` lookup on every row:

<img width="1620" height="307" alt="Image" src="https://github.com/user-attachments/assets/cccf1c69-be88-42dd-a788-d0c2f3f7184c" />

Assuming that we'll iterate over the `ResultSet` in a consistent fashion, always calling the same `ResultSet.getXyz(position)` method for any given position, we'll have to look it up only once for each column. Even if the `Xyz` type changes between rows, there aren't an infinite number of possible types, so we could cache each one for each column in a `Map<Class<?>, Converter<?, ?>>[]` data structure. Or probably better, assuming only one `Xyz` type is being looked up in 98% of the cases, a `Converter<?, ?>[]` data structure.

**Comment by lukaseder** [comment]
Yeah, this leads to a slight improvement as well:

```
Benchmark                  Mode  Cnt        Score      Error  Units
FetchBench.fetch           avgt   50  1443200.626 ± 6117.299  ns/op
FetchBench.fetchDirect     avgt   50    29026.989 ±  329.753  ns/op
FetchBench.fetchResultSet  avgt   50   284474.191 ± 2570.007  ns/op
```

Fixing this as:

- https://github.com/jOOQ/jOOQ/issues/19116

**Comment by lukaseder** [comment]
Closing this issue now.
